consensus/istanbul/backend/kip103.go

```diff
@@ -0,0 +1,179 @@
+	numRetiredBigInt, err := contract.GetRetiredCount(nil)
+	if err != nil {
+		logger.Error("Failed to get RetiredCount from TreasuryRebalance contract", "err", err)
+		return err
+	}
+
+	for i := 0; i < int(numRetiredBigInt.Int64()); i++ {
+		ret, err := contract.Retirees(nil, big.NewInt(int64(i)))
+		if err != nil {
+			logger.Error("Failed to get Retirees from TreasuryRebalance contract", "err", err)
+			return err
+		}
+		receipt.Retired[ret] = state.GetBalance(ret)
+	}
+	return nil
+}
+
+func (receipt *kip103Receipt) fillNewbie(contract *kip103.TreasuryRebalanceCaller) error {
+	numNewbieBigInt, err := contract.GetNewbieCount(nil)
+	if err != nil {
+		logger.Error("Failed to get NewbieCount from TreasuryRebalance contract", "err", err)
+		return nil
+	}
+
+	for i := 0; i < int(numNewbieBigInt.Int64()); i++ {
+		ret, err := contract.Newbies(nil, big.NewInt(int64(i)))
+		if err != nil {
+			logger.Error("Failed to get Newbies from TreasuryRebalance contract", "err", err)
+			return err
+		}
+		receipt.Newbie[ret.Newbie] = ret.Amount
+	}
+	return nil
+}
+
+func (receipt *kip103Receipt) totalRetriedBalance() *big.Int {
+	total := big.NewInt(0)
+	for _, bal := range receipt.Retired {
+		total.Add(total, bal)
+	}
+	return total
+}
+
+func (receipt *kip103Receipt) totalNewbieBalance() *big.Int {
+	total := big.NewInt(0)
+	for _, bal := range receipt.Newbie {
+		total.Add(total, bal)
+	}
+	return total
+}
+
+// rebalanceTreasury reads data from a contract, validates stored values, and executes treasury rebalancing (KIP-103).
+// It can change the global state by removing old treasury balances and allocating new treasury balances.
+// The new allocation can be larger than the removed amount, and the difference between two amounts will be burnt.
+func rebalanceTreasury(state *state.StateDB, chain consensus.ChainReader, header *types.Header) (*kip103Receipt, error) {
+	receipt := newKip103Receipt()
+	c := &kip103ContractCaller{state, chain, header}
+
+	// Inside check to avoid a panic case
+	if chain.Config().KIP103 == nil {
+		return receipt, errors.New("cannot find KIP103 configuration")
+	}
+
+	caller, err := kip103.NewTreasuryRebalanceCaller(chain.Config().KIP103.Kip103ContractAddress, c)
+	if err != nil {
+		return receipt, err
+	}
+
+	// Retrieve 1) Get Retired
+	if err := receipt.fillRetired(caller, state); err != nil {
+		return receipt, err
+	}
+
+	// Retrieve 2) Get Newbie
+	if err := receipt.fillNewbie(caller); err != nil {
+		return receipt, err
+	}
+
+	// Validation 1) Check the target block number
+	if blockNum, err := caller.RebalanceBlockNumber(nil); err != nil || blockNum.Cmp(header.Number) != 0 {
+		return receipt, errors.New("cannot find a proper target block number")
+	}
+
+	// Validation 2) Check whether status is approved. It should be 3 meaning approved
+	if status, err := caller.Status(nil); err != nil || status != 3 {
```

Comment:
Isn't Initialized=0, Registered=1, Approved=2, Finalized=3?
```suggestion
	if status, err := caller.Status(nil); err != nil || status != 2 {
```

Reply:
nice catch!